crates/k_means/src/approx.rs

```diff
@@ -0,0 +1,299 @@
+        |mut acc, (i, target)| {
+            acc[target].push(i);
+            acc
+        },
+    );
+    let alloc_size = alloc.iter().map(|x| x.len() as u32).collect::<Vec<_>>();
+    let keep_indices: Vec<usize> = alloc_size
+        .iter()
+        .enumerate()
+        .filter_map(|(i, size)| if *size > 0 { Some(i) } else { None })
+        .collect();
+    let alloc: Vec<_> = keep_indices.iter().map(|&i| alloc[i].clone()).collect();
+    let alloc_size: Vec<_> = keep_indices.iter().map(|&i| alloc_size[i]).collect();
+    let alloc_lists = successive_quotients_allocate(c, alloc_size);
+
+    let mut bottom_k_means = vec![];
+    let mut partition_start = vec![];
+    let mut offset = 0;
+    let all_sub_samples = partition_mut(samples, &alloc);
+    for (sub_samples, nlist) in all_sub_samples.into_iter().zip(alloc_lists) {
+        partition_start.push(offset);
+        offset += nlist as usize;
+        let f = crate::k_means(d, sub_samples, nlist as usize, num_threads, seed);
+        bottom_k_means.push(f);
+    }
+    Box::new(Approx {
+        this: This {
+            pool,
+            d,
+            c,
+            centroids,
+            targets,
+            rng,
+        },
+        top_centroids,
+        partition_start,
+        bottom_k_means,
+    })
+}
+
+pub fn k_means_dimension_restore<'a>(
+    d: usize,
+    c: usize,
+    num_points: usize,
+    is_spherical: bool,
+    index: Box<dyn Fn(&[f32]) -> u32 + 'a>,
+    mut next_sample: impl FnMut() -> Option<(Vec<f32>, Vec<f32>)>,
+) -> Square {
+    let mut centroids = vec![vec![0.0_f32; d]; c];
+    let mut ret = Square::new(d);
+    let mut count = vec![0_u32; c];
+    let mut traveled = 0;
+    while let Some((original, reduction)) = next_sample()
+        && traveled < num_points
+    {
+        let cid = index(&reduction) as usize;
+        centroids[cid] = f32::vector_add(&centroids[cid], &original);
+        count[cid] += 1;
+        traveled += 1;
+    }
+    (centroids, count) = centroids
+        .into_iter()
+        .zip(count)
+        .filter(|(_, cnt)| *cnt > 0)
+        .unzip();
+    for i in 0..count.len() {
+        k_means_centroids_inplace(&mut centroids[i], count[i], is_spherical);
+        ret.push_slice(&centroids[i]);
+    }
+    ret
+}
+
+fn k_means_assign(
+    samples: &mut SquareMut<'_>,
+    centroids: &Square,
+    num_threads: usize,
+) -> Vec<usize> {
+    use rayon::iter::{IndexedParallelIterator, IntoParallelRefMutIterator, ParallelIterator};
+    let n = samples.len();
+    let mut targets = vec![0; n];
+    let pool = rayon::ThreadPoolBuilder::new()
+        .num_threads(num_threads)
+        .build()
+        .expect("failed to build thread pool");
+    pool.install(|| {
+        targets
+            .par_iter_mut()
+            .zip(samples.par_iter_mut())
+            .for_each(|(target, sample)| {
+                *target = crate::flat::k_means_lookup(sample, centroids);
+            });
+    });
+    targets
+}
```

Comment:
This function is unnecessary, and it strangely creates the thread pool twice.

By the way, we should stop passing in `num_threads` and pass in `&ThreadPool` instead. Otherwise, this will create unnecessary threads. On my system, the maximum number of threads a process can create is 248,567 (`cat /proc/sys/kernel/threads-max`).